make_prg/interval_partition.py

```diff
@@ -0,0 +1,237 @@
+"""
+Code responsible for converting a consensus string into a set of disjoint
+match/non_match intervals.
+"""
+from enum import Enum, auto
+from typing import List, Tuple, Optional
+
+from make_prg import MSA
+
+from make_prg.seq_utils import get_interval_seqs, is_non_match, has_empty_sequence
+
+
+class PartitioningError(Exception):
+    pass
+
+
+class IntervalType(Enum):
+    Match = auto()
+    NonMatch = auto()
+
+
+def get_type(letter: str) -> IntervalType:
+    if letter == "*":
+        return IntervalType.NonMatch
+    else:
+        return IntervalType.Match
+
+
+def is_type(letter: str, interval_type: IntervalType) -> bool:
+    if get_type(letter) is interval_type:
+        return True
+    else:
+        return False
+
+
+class Interval:
+    def __init__(self, it_type: IntervalType, start: int, stop: int = None):
+        self.type = it_type
+        self.start = start
+        if stop is not None:
+            assert stop >= start
+        self.stop = stop if stop is not None else start
+
+    def modify_by(self, left_delta: int, right_delta: int):
+        self.start += left_delta
+        self.stop += right_delta
+
+    def contains(self, position: int):
+        return self.start <= position <= self.stop
```

Comment:
It might be nice to add a docstring mentioning what sort of intervals this class stores. i.e. half-open, open, closed. It seems like this interval is a closed interval?  
Also, just mentioning this incase you aren't aware of it: https://github.com/chaimleib/intervaltree  
I use this for all interval work and it is super nice. I'm not suggesting you replace the class here or anything. Just putting it on your radar if it isn't already in case there is need for it down the line.

Reply:
thanks, i added a docstring